recursion/gnark/server/load.go

```diff
@@ -0,0 +1,133 @@
+package server
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/feature/s3/manager"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/consensys/gnark-crypto/ecc"
+	"github.com/consensys/gnark/backend/groth16"
+	"github.com/consensys/gnark/constraint"
+	"github.com/pkg/errors"
+)
+
+// LoadCircuit downloads the R1CS and proving key and loads it into memory.
+func LoadCircuit(ctx context.Context, circuitBucket, circuitType, circuitVersion string) (constraint.ConstraintSystem, groth16.ProvingKey, error) {
+	// Setup AWS S3 downloader.
+	awsConfig, err := config.LoadDefaultConfig(ctx)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "loading AWS config")
+	}
+	s3Downloader := manager.NewDownloader(s3.NewFromConfig(awsConfig), func(d *manager.Downloader) {
+		d.PartSize = 128 * 1024 * 1024 // 128MB per part
+		d.Concurrency = 32
+	})
+
+	// Download the circuit tarball
+	tarballBuffer := manager.NewWriteAtBuffer(nil)
+	_, err = s3Downloader.Download(ctx, tarballBuffer, &s3.GetObjectInput{
+		Bucket: aws.String(circuitBucket),
+		Key:    aws.String(fmt.Sprintf("%s-build%s.tar.gz", circuitType, circuitVersion)),
+	})
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "downloading circuit")
+	}
+	tarballSize := len(tarballBuffer.Bytes())
+	if tarballSize == 0 {
+		return nil, nil, errors.New("downloaded tarball is empty")
+	}
+	fmt.Printf("Downloaded circuit tarball (%d bytes)\n", tarballSize)
+
+	// Decompress the gzip tarball.
+	gzipReader, err := gzip.NewReader(bytes.NewReader(tarballBuffer.Bytes()))
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "error decompressing gzip")
+	}
+
+	// Read the tarball.
+	tarReader := tar.NewReader(gzipReader)
+
+	// Temporary files to write and read to.
+	r1csTempFile, err := os.CreateTemp("", "r1cs")
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "creating temp file for r1cs")
+	}
+	defer os.Remove(r1csTempFile.Name())
+	pkTimeFile, err := os.CreateTemp("", "pk")
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "creating temp file for proving key")
+	}
+	defer os.Remove(r1csTempFile.Name())
+
+	r1csFileName := fmt.Sprintf("build/circuit_%s.bin", circuitType)
+	pkFileName := fmt.Sprintf("build/pk_%s.bin", circuitType)
+
+	var r1csExtracted, pkExtracted bool
+	for {
+		if r1csExtracted && pkExtracted {
+			break
```

Comment:
You should separate out downloading the artifacts to a folder with loading the circuit artifacts from a build directory. Right now this function does both.

Reply:
going to leave a TODO to refactor this since Uma wants to test